### DISCUSSION TITLE
API to select the partition on next reboot

### DISCUSSION THREAD

**Discussion by frankonski** [HomeSpan/HomeSpan#883] [Ideas]
Using `setRebootCallback()`, assuming a fresh re-flash to *app1* ended up being buggy and segfault'ing, it would be nice to leverage the reboot callback to force a reboot to the last known good partition being *app0* after 5 consecutive reboots within 30 seconds.

**Comment by HomeSpan** [comment]
The IDF has functionality designed to rollback for these reasons, but it's not implemented in the Arduino-ESP32 library.  I tried to find a good hook, but to no avail unless I edit the Arduino-ESP32 code directly (e.g. create my own Board Manager, which has some appeal for other reasons...).

The problem is that any reboot logic needs to be included BEFORE the `setup()` for it to be robust, else the logic may never get called if there is a reboot.  I'll check to see if the IDF feature designed for this has been enabled in version 3 of the Arduino-ESP32 library manager, but I'm still buried in trying to refactor the code to work in the new environment.

**Comment by frankonski** [comment]
Thanks about the intro on rollback. I found an easy-to-understand post [here](https://blog.espressif.com/ota-updates-framework-ab5438e30c12). 
HomeSpan doesn’t seem to call `esp_ota_mark_app_valid_cancel_rollback();` unless I didn’t find it, or this is the missing feature you are talking about.

But if that is the case, this means HomeSpan would keep the new partition in a _Pending Verify_ state which means the ability for anyone to manually rollback using `esp_ota_mark_app_invalid_rollback_and_reboot();` is still available.

I will do a quick test with `esp_ota_get_state_partition()` when I get a chance to test my assumption.

**Reply by HomeSpan** [reply]
HomeSpan will roll back if you upload, via OTA, a non-HomeSpan sketch, or a HomeSpan sketch that does not contain OTA. See the HomeSpan OTA docs for details on safe OTA loads.

**Reply by frankonski** [reply]
Yes, and this is one valid case for rollback, I agree.

**Comment by frankonski** [comment]
> The problem is that any reboot logic needs to be included BEFORE the setup() for it to be robust, else the logic may never get called if there is a reboot.

Not necessarily… there are lots of different use cases.. A segfault could occur in the main loop(), and by checking a reboot count + reboot reason + timing of these reboots, one could decide to rollback within the main loop().

**Comment by HomeSpan** [comment]
There is a formal a proper way to use rollbacks.  To do so requires a small change to the Arduino ESP board manager.  Probably makes the most sense to put in a request to the Espressif team to add that feature.  When I have some time I can explain in detail what and where the change is needed.

**Comment by frankonski** [comment]
I ended up experimenting a bit, and results seems promising. I can easily rollback to the previous version, and for fun I can switch back to the latest version uploaded OTA.

1. Define new UserSpan commands
```
  homeSpan.setSketchVersion("v1.0"); // change every build
  new SpanUserCommand('i', "Info on OTA", info_ota);
  new SpanUserCommand('r', "Revert OTA", revert_ota);
  new SpanUserCommand('s', "Switch back to latest OTA", switch_back_ota);
```

2. Implement the functions:
```
void info_ota(const char *buf){ 
  Serial.printf("Sketch version: %s\n", homeSpan.getSketchVersion());

  const esp_partition_t *lastInvalid = esp_ota_get_last_invalid_partition();
  esp_ota_img_states_t ota_state;


  Serial.printf("Partition running: %s\n", esp_ota_get_running_partition()->label);
  Serial.printf("Last invalid partition: %s\n", lastInvalid ? lastInvalid->label : "None");
  Serial.printf("Is rollback possible? %s\n", esp_ota_check_rollback_is_possible() ? "Yes" : "No");

  if (esp_ota_get_state_partition(esp_ota_get_running_partition(), &ota_state) == ESP_OK) {
    switch (ota_state){
      case ESP_OTA_IMG_NEW:
        Serial.printf("State: ESP_OTA_IMG_NEW\n");
        break;
      case ESP_OTA_IMG_PENDING_VERIFY:
        Serial.printf("State: ESP_OTA_IMG_PENDING_VERIFY\n");
        break;
      case ESP_OTA_IMG_VALID:
        Serial.printf("State: ESP_OTA_IMG_VALID\n");
        break;
      case ESP_OTA_IMG_INVALID:
        Serial.printf("State: ESP_OTA_IMG_INVALID\n");
        break;
      case ESP_OTA_IMG_ABORTED:
        Serial.printf("State: ESP_OTA_IMG_ABORTED\n");
        break;
      case ESP_OTA_IMG_UNDEFINED:
        Serial.printf("State: ESP_OTA_IMG_UNDEFINED\n");
        break;
    }
  } else {
    Serial.printf("Error getting state partition\n");
  }
};

void revert_ota(const char *buf){ 

    Serial.printf("Forcing rollback ...\n");
    esp_ota_mark_app_invalid_rollback_and_reboot();

};

// Calling this method currently assumed a rollback was already done. So switching back to the latest OTA version flashed.
void switch_back_ota(const char *buf){ 

  const esp_partition_t *lastUploaded = esp_ota_get_last_invalid_partition();

  if (esp_ota_set_boot_partition(lastUploaded) == ESP_OK) {
    Serial.printf("Switch back to recent version ...\n");
    esp_restart();
  }

};
```

Upload a first version of the code, check the sketch version and details using `@i`.
Change the sketch version.
Build and upload a second version of the sketch.
From version `2.0` you will then be able to revert back to 1.0, and switch back to reboot back in 2.0.

This means I could leverage that to implement some kind of "automatic revert when reboot reason is Panic if 5 reboots in 30 seconds".

**Comment by frankonski** [comment]
I noticed this commit: https://github.com/HomeSpan/HomeSpan/commit/01028937f9503f7dfef92a1e36a04b297694df32
Not a bad idea how you are thinking about it.
